**Linux with security manager** [Build 4500](https://ci.wildfly.org/viewLog.html?buildId=18132&buildTypeId=WF_PullRequest_LinuxSm) outcome was **FAILURE** using a merge of 2c3f4d660f5c918c7afdd7ef91c68d8b2ae1cd23
Summary: Tests failed: 1, passed: 1004, ignored: 121 Build time: 00:10:42
### Failed tests
```
org.jboss.as.test.integration.web.handlers.UndertowHandlersConfigTestCase.testRewrite: java.lang.AssertionError: expected:<200> but was:<404>
	at org.jboss.as.test.integration.web.handlers.UndertowHandlersConfigTestCase.testRewrite(UndertowHandlersConfigTestCase.java:56)
------- Stdout: -------
 [0m05:58:56,244 INFO  [org.wildfly.extension.undertow] (ServerService Thread Pool -- 4) WFLYUT0022: Unregistered web context: /form-auth
 [0m [0m05:58:56,267 INFO  [org.jboss.as.server.deployment] (MSC service thread 1-7) WFLYSRV0028: Stopped deployment form-auth.war (runtime-name: form-auth.war) in 25ms
 [0m [0m05:58:56,278 INFO  [org.jboss.as.repository] (management-handler-thread - 1) WFLYDR0002: Content removed from location /opt/buildAgent/work/a31d203e70e89f90/testsuite/integration/web/target/jbossas/standalone/data/content/03/0ff77b121b3cf37cd54022f2f623bb7e53dbf9/content
 [0m [0m05:58:56,278 INFO  [org.jboss.as.server] (management-handler-thread - 1) WFLYSRV0009: Undeployed "form-auth.war" (runtime-name: "form-auth.war")
 [0m [0m05:58:56,315 INFO  [org.jboss.as.repository] (management-handler-thread - 3) WFLYDR0001: Content added at location /opt/buildAgent/work/a31d203e70e89f90/testsuite/integration/web/target/jbossas/standalone/data/content/0b/81e197406d477fbd2e163bd1cd37a34668fd94/content
 [0m [0m05:58:56,320 INFO  [org.jboss.as.server.deployment] (MSC service thread 1-7) WFLYSRV0027: Starting deployment of "non-blocking-handler.war" (runtime-name: "non-blocking-handler.war")
 [0m [33m05:58:56,369 WARN  [org.jboss.as.dependency.private] (MSC service thread 1-7) WFLYSRV0018: Deployment "deployment.non-blocking-handler.war" is using a private module ("org.wildfly.security.manager:main") which may be changed or removed in future versions without notice.
 [0m [33m05:58:56,379 WARN  [io.undertow] (MSC service thread 1-7) UT005060: Predicate regex['/rewrite.*'] -> rewrite['/file.txt'] uses old style square braces to define predicates, which will be removed in a future release. predicate[value] should be changed to predicate(value)
 [0m [33m05:58:56,381 WARN  [io.undertow] (MSC service thread 1-7) UT005060: Predicate regex['/rewrite.*'] -> rewrite['/file.txt'] uses old style square braces to define predicates, which will be removed in a future release. predicate[value] should be changed to predicate(value)
 [0m [0m05:58:56,428 INFO  [org.jboss.as.arquillian] (MSC service thread 1-7) Arquillian deployment detected: ArquillianConfig[service=jboss.arquillian.config."non-blocking-handler.war",unit=non-blocking-handler.war,tests=[org.jboss.as.test.integration.web.handlers.UndertowHandlersConfigTestCase, org.jboss.as.test.integration.web.handlers.RequestDumpingHandlerTestCase]]
 [0m [0m05:58:56,468 INFO  [org.wildfly.extension.undertow] (ServerService Thread Pool -- 8) WFLYUT0021: Registered web context: /non-blocking-handler
 [0m [0m05:58:56,477 INFO  [org.jboss.as.server] (management-handler-thread - 3) WFLYSRV0010: Deployed "non-blocking-handler.war" (runtime-name : "non-blocking-handler.war")
 [0m


```

--- a/testsuite/integration/web/src/test/java/org/jboss/as/test/integration/web/handlers/UndertowHandlersConfigTestCase.java
+++ b/testsuite/integration/web/src/test/java/org/jboss/as/test/integration/web/handlers/UndertowHandlersConfigTestCase.java
@@ -11,6 +11,7 @@ import org.jboss.arquillian.container.test.api.Deployment;
 import org.jboss.arquillian.container.test.api.RunAsClient;
 import org.jboss.arquillian.junit.Arquillian;
 import org.jboss.arquillian.test.api.ArquillianResource;
+import org.jboss.as.test.shared.integration.ejb.security.PermissionUtils;
 import org.jboss.shrinkwrap.api.ShrinkWrap;
 import org.jboss.shrinkwrap.api.asset.StringAsset;
 import org.jboss.shrinkwrap.api.spec.WebArchive;
@@ -18,6 +19,7 @@ import org.junit.Assert;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 
+import java.io.FilePermission;
 import java.net.URL;
 
 import static org.junit.Assert.assertEquals;
@@ -37,7 +39,9 @@ public class UndertowHandlersConfigTestCase {
                 .addPackage(UndertowHandlersConfigTestCase.class.getPackage())
                 .addAsWebInfResource(UndertowHandlersConfigTestCase.class.getPackage(), "jboss-web.xml", "jboss-web.xml")
                 .addAsWebResource(new StringAsset("A file"), "file.txt")
-                .addAsWebInfResource(new StringAsset("regex['/rewrite.*'] -> rewrite['/file.txt']"), "undertow-handlers.conf");
+                .addAsWebInfResource(new StringAsset("regex['/rewrite.*'] -> rewrite['/file.txt']"), "undertow-handlers.conf")
+                .addAsWebResource(PermissionUtils.createPermissionsXmlAsset(new FilePermission("<<ALL FILES>>", "read,write")), "META-INF/permissions.xml");
+
     }
 
     @ArquillianResource

--- a/undertow/src/main/java/org/wildfly/extension/undertow/deployment/RewriteCorrectingHandlerWrappers.java
+++ b/undertow/src/main/java/org/wildfly/extension/undertow/deployment/RewriteCorrectingHandlerWrappers.java
@@ -1,0 +1,76 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2016, Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.wildfly.extension.undertow.deployment;
+
+import io.undertow.server.HandlerWrapper;
+import io.undertow.server.HttpHandler;
+import io.undertow.server.HttpServerExchange;
+import io.undertow.servlet.handlers.ServletPathMatch;
+import io.undertow.servlet.handlers.ServletRequestContext;
+import io.undertow.util.AttachmentKey;
+
+/**
+ * Handler that works around issues with rewrites() and undertow-handlers.conf.
+ * <p>
+ * Because the rewrite happens after the initial dispatch this handler detects if
+ * the path has been rewritten and updates the servlet target.
+ *
+ * This is a bit of a hack, it needs a lot more thinking about a clean way to handle
+ * this
+ */
+class RewriteCorrectingHandlerWrappers {
+
+    private static final AttachmentKey<String> OLD_RELATIVE_PATH = AttachmentKey.create(String.class);
+
+    static class PreWrapper implements HandlerWrapper {
+
+        @Override
+        public HttpHandler wrap(final HttpHandler handler) {
+            return new HttpHandler() {
+                @Override
+                public void handleRequest(HttpServerExchange exchange) throws Exception {
+                    exchange.putAttachment(OLD_RELATIVE_PATH, exchange.getRelativePath());
+                    handler.handleRequest(exchange);
+                }
+            };
+        }
+    }
+
+    static class PostWrapper implements HandlerWrapper {
+        @Override
+        public HttpHandler wrap(final HttpHandler handler) {
+            return new HttpHandler() {
+                @Override
+                public void handleRequest(HttpServerExchange exchange) throws Exception {
+                    String old = exchange.getAttachment(OLD_RELATIVE_PATH);
+                    if(!old.equals(exchange.getRelativePath())) {
+                        ServletRequestContext src = exchange.getAttachment(ServletRequestContext.ATTACHMENT_KEY);
+                        ServletPathMatch info = src.getDeployment().getServletPaths().getServletHandlerByPath(exchange.getRelativePath());
+                        src.setCurrentServlet(info.getServletChain());
+                        src.setServletPathMatch(info);
+                    }
+                    handler.handleRequest(exchange);
+                }
+            };
+        }
+    }
+}

--- a/undertow/src/main/java/org/wildfly/extension/undertow/deployment/UndertowDeploymentInfoService.java
+++ b/undertow/src/main/java/org/wildfly/extension/undertow/deployment/UndertowDeploymentInfoService.java
@@ -1043,7 +1043,8 @@ public class UndertowDeploymentInfoService implements Service<DeploymentInfo> {
             }
 
             if (predicatedHandlers != null && !predicatedHandlers.isEmpty()) {
-                d.addInitialHandlerChainWrapper(new HandlerWrapper() {
+                d.addOuterHandlerChainWrapper(new RewriteCorrectingHandlerWrappers.PostWrapper());
+                d.addOuterHandlerChainWrapper(new HandlerWrapper() {
                     @Override
                     public HttpHandler wrap(HttpHandler handler) {
                         if (predicatedHandlers.size() == 1) {
@@ -1054,6 +1055,7 @@ public class UndertowDeploymentInfoService implements Service<DeploymentInfo> {
                         }
                     }
                 });
+                d.addOuterHandlerChainWrapper(new RewriteCorrectingHandlerWrappers.PreWrapper());
             }
 
             if (mergedMetaData.getDefaultEncoding() != null) {

